src/main/java/com/google/sps/algorithms/MatchRanking.java

```diff
@@ -0,0 +1,121 @@
+    */
+    private static Map<String, Double> scoreNewMatches(Collection<String> savedMatchBios, Collection<String> allUserBios, Collection<String> newMatchBios) {
+        //count words in all of the saved matches
+        Map<String, Integer> savedMatchesWordCount = countWordInstances(savedMatchBios);
+
+        //count words in all of the user bios
+        Map<String, Integer> allUserWordCount = countWordInstances(allUserBios);
+
+        //calculate score for each new match bio
+        Map<String, Double> newBioScores = new HashMap<String, Double>();
+        for(String newBio : newMatchBios) {
+            double bioScore = calculateBioScore(newBio, savedMatchesWordCount, allUserWordCount);
+            newBioScores.put(newBio, new Double(bioScore));
+        }
+
+        return newBioScores;
+    }
+
+    /**
+    * Function that takes the map of bios and their scores and returns a list of the bios in sorted order from highest to lowest score.
+    */
+    private static List<String> sortBiosByScore(Map<String, Double> newMatchScores) {
+        List<String> orderedBios = new ArrayList<String>();
+        Object[] scoresArray = newMatchScores.entrySet().toArray();
+
+        Arrays.sort(scoresArray, new Comparator() {
+            public int compare(Object o1, Object o2) {
+                return ((Map.Entry<String, Double>) o2).getValue()
+                        .compareTo(((Map.Entry<String, Double>) o1).getValue());
+            }
+        });
+
+        for (Object entry : scoresArray) {
+        orderedBios.add(((Map.Entry<String, Double>) entry).getKey());
+        }
+
+        return orderedBios;
+    }
+
+    /**
+    * Function that takes a bio and maps of the word counts in saved bios and all bios, and returns the score for the 
+    * bio as a double.
+    */
+    private static double calculateBioScore(String bio, Map<String, Integer> savedMatchesWordCount, Map<String, Integer> allUserWordCount) {
+        //separate bio words by whitespace and store in an array
+        String[] bioWords = bio.toLowerCase().split("\\W+");
+
+        double numerator = calculateBioProbability(bioWords, savedMatchesWordCount);
+        double denominator = calculateBioProbability(bioWords, allUserWordCount);
+
+        return (numerator/denominator);
+    }
+
+    /**
+    * Calculate the probability of generating a bio given a certain word count map.
+    */
+    private static double calculateBioProbability(String[] bioWords, Map<String, Integer> givenMap) {
+        double toReturn = 0.0;
+
+        for(String word : bioWords) {
+            Integer wordCount = givenMap.get(word);
+
+            //skip word if it is not in the given map
+            if(wordCount != null) {
+                double wordCountDouble = wordCount.intValue();
+                //take log to make large numbers easier to analyze
+                toReturn += Math.log(wordCountDouble / givenMap.size());
+            }
+        }
+
+        return toReturn;
```

Comment:
Could you convert this back to a probability from a log probability? Or change the method name and comments to reflect this is a log-prob (and then fix the numerator / denominator division).

Reply:
I took out the log probability because the nth root can't take negatives, and the log probabilities < 1 were negative

Reply:
This is fine for now (but note you may run into numerical precision errors again due to repeated multiplying as you saw before --  Java's smallest double is 2^-1074, which you'd reach with, for instance, ~100 words with frequency 1/1000).

The other way of addressing this was to return log probs (as before), then instead of doing (numerator_prob / denominator_prob) ^(1/n) to get the score, you'd do the same but using the logarithmic equivalents:
log(score)
= log((numerator_prob / denominator_prob) ^(1/n))
= (log_numerator - log_denominator)/n